File: data/java/tests/test_shutil_rmtree.java
package com.example;

import org.junit.jupiter.api.*;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;


import static org.junit.jupiter.api.Assertions.*;

public class test_shutil_rmtree {
    private function_shutil_rmtree deleter;
    private String testDir;

    @BeforeEach
    void setUp() throws IOException {
        deleter = new function_shutil_rmtree();
        testDir = "test_directory";
        createTestDirectory(testDir);
    }

    @AfterEach
    void tearDown() {
        File dir = new File(testDir);
        if (dir.exists()) {
            cleanup(testDir);
        } else {
            System.out.printf("Directory '%s' does not exist, skipping cleanup.%n", testDir);
        }
    }

    private void createTestDirectory(String directoryPath) throws IOException {
        File directory = new File(directoryPath);
        directory.mkdirs();

        try (FileWriter fw = new FileWriter(new File(directory, "file1.txt"))) {
            fw.write("This is a test file.");
        }

        File subDir = new File(directory, "subdir");
        subDir.mkdirs();

        try (FileWriter fw = new FileWriter(new File(subDir, "file2.txt"))) {
            fw.write("This is another test file.");
        }
    }

    private void cleanup(String directoryPath) {
        String result = deleter.delete_directory_tree(directoryPath);
        if (!result.contains("deleted")) {
            throw new RuntimeException("Failed to clean up directory: " + result);
        }
    }

    @Test
    void testDeleteExistingDirectory() {
        String result = deleter.delete_directory_tree(testDir);
        assertTrue(result.startsWith(String.format("Directory '%s' deleted successfully.", testDir)));
    }

    @Test
    void testDeleteNonExistentDirectory() {
        String nonExistentDir = "non_existent_directory";
        String result = deleter.delete_directory_tree(nonExistentDir);
        assertEquals(String.format("Directory '%s' does not exist.", nonExistentDir), result);
    }

    @Test
    void testDeleteDirectoryWithFiles() {
        String result = deleter.delete_directory_tree(testDir);
        assertTrue(result.startsWith(String.format("Directory '%s' deleted successfully.", testDir)));
    }

    @Test
    void testDeleteWithPermissionsError() throws IOException {
        File protectedDir = new File(testDir, "protected");
        protectedDir.mkdirs();
        File protectedFile = new File(protectedDir, "protected_file.txt");

        try (FileWriter fw = new FileWriter(protectedFile)) {
            fw.write("Protected content.");
        }

        // 将文件设置为只读
        protectedFile.setReadOnly();
        // 确保目录也是只读的
        protectedDir.setReadOnly();

        String result = deleter.delete_directory_tree(protectedDir.getPath());

        // 即使文件是只读的，我们现在也应该能够删除它
        assertTrue(result.contains("Failed to delete directory") || !new File(protectedDir.getPath()).exists());
    }

    @Test
    void testPartialDeletion() {
        File protectedFile = new File(testDir, "protected_file.txt");
        String result = "";

        try {
            try (FileWriter fw = new FileWriter(protectedFile)) {
                fw.write("Partial deletion test.");
            }

            // 将文件设置为只读
            protectedFile.setReadOnly();

            result = deleter.delete_directory_tree(testDir);
        } catch (IOException e) {
            result = e.getMessage();
        }

        // 现在我们应该能够正确处理权限问题
        assertTrue(result.contains("Failed to delete directory") || result.contains("Permission denied") ||
                !new File(testDir).exists());
    }
}

File: data/cpp/tests/test_shutil_rmtree.cpp
#include <gtest/gtest.h>

#include "../src/function_shutil_rmtree.cpp"

class TestShutilRmtree : public ::testing::Test {
protected:
    DirectoryDeleter deleter;
    std::string test_dir = "test_directory";

    void SetUp() override {
        create_test_directory(test_dir);
    }

    void TearDown() override {
        if (std::filesystem::exists(test_dir)) {
            std::filesystem::remove_all(test_dir);
        }
    }

    void create_test_directory(const std::string& directory_path) {
        std::filesystem::create_directories(directory_path);
        
        std::ofstream file1(directory_path + "/file1.txt");
        file1 << "This is a test file.";
        file1.close();

        std::string sub_dir = directory_path + "/subdir";
        std::filesystem::create_directories(sub_dir);
        
        std::ofstream file2(sub_dir + "/file2.txt");
        file2 << "This is another test file.";
        file2.close();
    }
};

TEST_F(TestShutilRmtree, TestDeleteExistingDirectory) {
    std::string result = deleter.delete_directory_tree(test_dir);
    EXPECT_TRUE(result.find("deleted successfully") != std::string::npos);
}

TEST_F(TestShutilRmtree, TestDeleteNonExistentDirectory) {
    std::string non_existent_dir = "non_existent_directory";
    std::string result = deleter.delete_directory_tree(non_existent_dir);
    EXPECT_EQ(result, "Directory '" + non_existent_dir + "' does not exist.");
}

TEST_F(TestShutilRmtree, TestDeleteDirectoryWithFiles) {
    std::string result = deleter.delete_directory_tree(test_dir);
    EXPECT_TRUE(result.find("deleted successfully") != std::string::npos);
}

TEST_F(TestShutilRmtree, TestDeleteWithPermissionsError) {
    #ifdef _WIN32
        // 在 Windows 上创建测试目录和只读文件
        std::string protected_dir = test_dir + "/protected";
        std::filesystem::create_directories(protected_dir);
        std::string protected_file = protected_dir + "/protected_file.txt";
    
        {
            std::ofstream file(protected_file);
            file << "Protected content.";
        }
    
        // 如果文件存在，则将其设置为“只读”（模拟无权限）
        if (std::filesystem::exists(protected_file)) {
            std::string cmd = "attrib +R \"" + protected_file + "\"";
            std::system(cmd.c_str());
        }
    
        // 调用删除目录操作
        std::string result = deleter.delete_directory_tree(protected_dir);
        EXPECT_TRUE(result.find("Failed to delete directory") != std::string::npos)
            << "Result was: " << result;
    
        // 如果文件仍然存在，则将其属性恢复为可写
        if (std::filesystem::exists(protected_file)) {
            std::string cmd = "attrib -R \"" + protected_file + "\"";
            std::system(cmd.c_str());
        }
    #else
        GTEST_SKIP() << "Skipping Windows-specific permissions test on non-Windows OS.";
    #endif
    }
    
    TEST_F(TestShutilRmtree, TestPartialDeletion) {
    #ifdef _WIN32
        // 创建只读文件
        std::string protected_file = test_dir + "/protected_file.txt";
        {
            std::ofstream file(protected_file);
            file << "Partial deletion test.";
        }
    
        // 设置为只读
        if (std::filesystem::exists(protected_file)) {
            std::string cmd = "attrib +R \"" + protected_file + "\"";
            std::system(cmd.c_str());
        }
    
        // 调用删除目录操作
        std::string result = deleter.delete_directory_tree(test_dir);
        EXPECT_TRUE(result.find("Failed to delete directory") != std::string::npos)
            << "Result was: " << result;
    
        // 如果文件仍然存在，则恢复其可写属性，方便后续清理
        if (std::filesystem::exists(protected_file)) {
            std::string cmd = "attrib -R \"" + protected_file + "\"";
            std::system(cmd.c_str());
        }
    #else
        GTEST_SKIP() << "Skipping Windows-specific permissions test on non-Windows OS.";
    #endif
    }
int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}